andi/andi.py

```diff
@@ -24,15 +26,19 @@ def inspect(func: Callable) -> Dict[str, List[Optional[Type]]]:
     return res
 
 
+TypeContainerOrCallable = Union[
+    Container[Type],
+    Callable[[Optional[Type]], bool]
+]
+
+
 def to_provide(
         arguments_or_func: Union[
             Callable,
             Dict[str, List[Optional[Type]]]
         ],
-        can_provide: Union[
-            Container[Type],
-            Callable[[Optional[Type]], bool]
-        ]
+        can_provide: TypeContainerOrCallable,
+        bindings: Callable[[Type], Optional[Type]] = None,
```

Comment:
hey @ivanprado! Would you mind if we move bindings to a separate PR, to simplify the review? If I'm not mistaken, this is an unrelated feature, and more like "nice to have", as this logic can be handled in the dependencies themselves. Having bindings would require more work (e.g. docstring and README updates, probably a naming discussion as well).

Reply:
done in https://github.com/scrapinghub/andi/pull/8

Reply:
Yes, no problem to discuss this in a different PR. 

> as this logic can be handled in the dependencies themselves

How can this logic be handled in the dependencies itself? Let's imagine a case that bindings solve:

```py
class Breadcrumbs(ABC):

    @abstractmethod
    def urls():
        pass

class ScrapyBreadcrumbs(WebPage, Breadcrumbs):

    def urls():
        return self.css("...").getall()

@attr.s(auto_attrib=True)
class AutoextractBreadcrumbs(Injectable, Breadcrumbs):
    autoextract_resp: AutoextractProductResponse

    def urls():
        return urls_from(autoextract_resp['breadcrumbs'])

# So there are two possible implementations for Breadcrumbs, but each
# has different dependencies (one depends on ResponseData, and the other on 
# AutoextractProductResponse). Andi could resolve them properly with bindings.

def parse(response, breadcrumbs: Breadcrumbs):
    pass

# This one resolves properly dependencies using Scrapy breadcrumbs version
andi.plan(parse, can_provide=..., externally_provided=..., 
          bindings={Breadcrumbs: ScrapyBreadcrumbs}.get)

# This one resolves properly dependencies using AutoExtract breadcrumbs version
andi.plan(parse, can_provide=..., externally_provided=..., 
          bindings={Breadcrumbs: AutoextractBreadcrumbs}.get)
```

Reply:
I see, I haven't considered that they could have different dependencies; makes sense.

Reply:
Separated PR created for that: https://github.com/scrapinghub/andi/pull/9